File: lib/bot/eval.go
package bot

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/go-resty/resty/v2"
	log "github.com/sirupsen/logrus"
	tgb "gopkg.in/telebot.v4"
)

func Eval(ctx tgb.Context) error {
	_, src, _ := strings.Cut(ctx.Message().Text, " ")
	if strings.TrimSpace(src) == "" {
		return ctx.Reply("Usage: /eval <rust_code>")
	}

	wrappedSrc := wrapMain(src)
	log.Debugf("/eval: evaluating the following snippet:\n```rs\n%s```", wrappedSrc)

	endpoint := "https://play.rust-lang.org/execute"
	body := map[string]any{
		"crateType": "bin",
		"channel":   "nightly",
		"edition":   "2024",
		"mode":      "debug",
		"tests":     false,
		"backtrace": true,
		"code":      wrappedSrc,
	}

	resp, err := resty.New().R().SetBody(body).Post(endpoint)
	if err != nil {
		return err
	}

	var evalRes rustPlaygroundResp
	if err := json.Unmarshal(resp.Body(), &evalRes); err != nil {
		return err
	}

	leader := ":)"
	if !evalRes.Success {
		leader = ":<"
		if evalRes.Error != "" {
			return ctx.Reply(fmt.Sprintf("%s %s", leader, evalRes.Error))
		}
	}
	return ctx.Reply(fmt.Sprintf(
		"%s %s\n\nSTDOUT\n%s\nSTDERR\n%s",
		leader, evalRes.ExitDetail, evalRes.Stdout, evalRes.Stderr,
	))
}

type rustPlaygroundResp struct {
	// The error message emitted by the playground server, if any.
	Error          string
	Success        bool
	ExitDetail     string
	Stdout, Stderr string
}

func wrapMain(src string) string {
	var blk string
	switch {
	case
		strings.Contains(src, "print!("),
		strings.Contains(src, "println!("):
		blk = "{\n%s\n};"
	default:
		blk = "println!(\"{:?}\", {\n%s\n});"
	}
	return fmt.Sprintf(
		`#[allow(warnings)] fn main() -> Result<(), Box<dyn std::error::Error>> {
%s
	Ok(())
}`,
		fmt.Sprintf(blk, src),
	)
}

File: lib/angela/command/eval.ex
defmodule Angela.Command.Eval do
  @moduledoc """
  Evaluates Rust code using the Rust Playground.
  """

  alias Angela.Command.Response
  alias ExGram.Model.{Message, ReplyParameters}

  @behaviour Angela.Command

  @impl true
  def usage(), do: "/eval <rust_code>"

  @impl true
  def respond(msg = %Message{text: src}) when src != "" do
    %{
      "crateType" => "bin",
      "channel" => "nightly",
      "edition" => "2024",
      "mode" => "debug",
      "tests" => false,
      "backtrace" => true,
      "code" => wrap_main(src)
    }
    |> query()
    |> case do
      {:ok, res} -> msg(res)
      {:error, reason} -> "error: #{inspect(reason)}"
    end
    |> Response.new(reply_parameters: %ReplyParameters{message_id: msg.message_id})
  end

  defp msg(result) do
    leader = if result["success"], do: ":)", else: ":<"

    case result do
      %{"success" => false, "error" => error} when error != "" ->
        "#{leader} #{error}"

      %{"exitDetail" => exit_detail, "stdout" => stdout, "stderr" => stderr} ->
        """
        #{leader} #{exit_detail}

        STDOUT
        #{stdout}
        STDERR
        #{stderr}\
        """

      _ ->
        "#{leader} unknown response format"
    end
  end

  @api_endpoint "https://play.rust-lang.org/execute"
  defp query(body) do
    case Tesla.post(client(), @api_endpoint, body) do
      {:ok, %Tesla.Env{status: 200, body: body}} -> {:ok, body}
      {:ok, %Tesla.Env{status: status}} -> {:error, "HTTP #{status}"}
      {:error, reason} -> {:error, reason}
    end
  end

  defp client() do
    [
      {Tesla.Middleware.Headers, [{"content-type", "application/json"}]},
      Tesla.Middleware.JSON
    ]
    |> Tesla.client()
  end

  defp wrap_main(src) do
    block =
      if src =~ ~r/print(ln)?!\(/ do
        "{\n#{src}\n};"
      else
        "println!(\"{:?}\", {\n#{src}\n});"
      end

    """
    #[allow(warnings)] fn main() -> Result<(), Box<dyn std::error::Error>> {
    #{block}
    \tOk(())
    }\
    """
  end
end
